### ISSUE TITLE
Considering iterable as proper return type declaration for generators

### ISSUE THREAD

**Issue by dnaber-de** [inpsyde/php-coding-standards#38]
`Inpsyde.CodeQuality.ReturnTypeDeclaration.NoGeneratorReturnType` complains about the following:

```
public function listSomething(): iterable {
     yield 1;
     yield 2;
}
```

I'm not sure if `iterable` suggests a rewinding of the returned list because that wouldn't be possible. If not, this construct should be allowed IMO.

**Comment by gmazzap** [comment]
The error exists because when the rule was written `iterable` was not a valid return type. So yes, there should no error for this.

That said, I would not suggest to use `iterable` as return type for a `Generator`.

`iterable` means there's no guarantee the return value is an object, so the only way to _iterate_ is to use `foreach`. 

Which means there's no way to check if the iterable is consumed. Take as example the code:

```php
function loopTheList(iterable $iterable): void
{
     // don't know if is an object, I can't call methods... can only use foreach
     foreach ($iterable as $item) {
        //
     }

     thenCallThis($iterable);
}


function thenCallThis(iterable $iterable): void
{
     // don't know if is an object, I can't call methods... can only use foreach
     foreach ($iterable as $item) {
        //
     }
}


loopTheList(listSomething());
```

I've used the functions according to their signature, but still a fatal error occurs.

So, unless we want to have a `try` / `catch` block everytime we use a `foreach` on something is not an array, this is not ideal.

So even if generators satisfy `iterable` type, I would suggest, when possible, to use at least `Iterator`, so that clients of the return value could use `while ($it->isValid()) { $it->next(); }` to iterate, and that is compatible with Generators. Still even `Iterator` is not ideal... because `rewind()` method could cause an exception, but using `try` / `catch` to wrap a `rewind()` is not the same to use it for every `foreach`.

Anyway, I'm going to fix this rule because is signaling an error that is not an error.

**Comment by gmazzap** [comment]
8b3abeb fixes this.

**Comment by dnaber-de** [comment]
> That said, I would not suggest to use iterable as return type for a Generator.

Good point

**Comment by gmazzap** [comment]
Included in [1.0.0-beta.7](https://github.com/inpsyde/php-coding-standards/releases/tag/1.0.0-beta.7).

Closing. Thanks for reporting.
